### DISCUSSION TITLE
File distribution policy/configuration help (mspmfs?)

### DISCUSSION THREAD

**Discussion by red217** [trapexit/mergerfs#1168] [Q&A]
Hello!

I recently discovered mergerfs and, since it fits my use case perfectly, I tried setting it up. It's great software and I managed to get it working, but I haven't been able to configure it to distribute new files as I would like. I read the documentation and believed I had found the options I needed, but it seems I was wrong. I want to keep this as short as possible, so I'll just describe my setup and how I'd like it to work.

First, I am running **mergerfs 2.31.0**, from the Debian stable repos.
I have 4 drives with their mount points added as branches. 3 of them are 8TB drives and contain many files from before I was using mergerfs, the 4th drive is a new 10TB drive which was new and empty when added. Since the first 3 drives already contain a manually organized file structure, I would like mergerfs to add new files to branches 1 to 3 if those paths already exist there, and if a path doesn't exist or a drive fills up it can go to branch 4, on the new and empty drive. For this reason, I chose `mspmfs` as the create policy, as I understand it will choose the branch which contains the path, before going up a level and trying again if it doesn't find a match.

After reading the documentation and following the basic setup recommendation, I ended up with this in fstab:
`/mnt/bank/backup[1-4]   /mnt/merge/backup       fuse.mergerfs allow_other,use_ino,cache.files=partial,dropcacheonclose=true,category.create=mspmfs,moveonenospc=true  0       0`

The mount works, but new files aren't being distributed as I expected. For instance I created a new directory through mergerfs under a path which only exists on branch 1 (on `/mnt/bank/backup1`) and expected to find it under `backup1`, but instead the whole missing path tree was created on `backup4` as well, which again was a new, empty drive. There were still over 2TB free on backup1, so it didn't run out of space. I don't understand why it was created on branch 4 and not branch 1 in this case.

What am I doing wrong or misunderstanding here? As I understand, mspmfs should get a match on `backup1` and select that branch before it can find a match on `backup4` by walking up the tree, but instead it seems it selects `backup4` anyway.

**Comment by trapexit** [comment]
Without you providing me very precise information about what is there and what happened one step at a time I have a hard time helping here. mspmfs works like described in the docs. It considers only the branches with the relative base path to the file to be created and if all branches are filtered out it walks back one level and tries again. If you're claiming this isn't happening then please provide a thorough step by step to reproduce. I just tested it myself and looks to work as expected.

**Reply by red217** [reply]
Thank you for the reply.

I will provide a precise, step by step example of the behavior I am seeing, as a sequence of commands and their output. I think that should be more clear than me trying to describe everything in English sentences. The mergerfs mount is as described by the fstab entry I already posted.

As an example, let's say I want to create a new directory called `TEST_DIRECTORY` using mergerfs, under the existing path `/mnt/merge/backup/raid-backup/data1/Games/GOG`
This is what I'm seeing:
```
$ ls -d /mnt/merge/backup/raid-backup/data1/Games/GOG
/mnt/merge/backup/raid-backup/data1/Games/GOG
```
The `GOG` directory exists under mergerfs, as it should. Now looking at each branch:
```
$ ls -d /mnt/bank/backup1/raid-backup/data1/Games/GOG
/mnt/bank/backup1/raid-backup/data1/Games/GOG

$ ls -d /mnt/bank/backup2/raid-backup/data1/Games/GOG
ls: cannot access '/mnt/bank/backup2/raid-backup/data1/Games/GOG': No such file or directory

$ ls -d /mnt/bank/backup3/raid-backup/data1/Games/GOG
ls: cannot access '/mnt/bank/backup3/raid-backup/data1/Games/GOG': No such file or directory

$ ls -d /mnt/bank/backup4/raid-backup/data1/Games/GOG
ls: cannot access '/mnt/bank/backup4/raid-backup/data1/Games/GOG': No such file or directory
```
So the `GOG` directory only exists under `/mnt/bank/backup1` and not on any other branch. Now to create the new directory via mergerfs:
```
$ mkdir /mnt/merge/backup/raid-backup/data1/Games/GOG/TEST_DIRECTORY

$ ls -d /mnt/merge/backup/raid-backup/data1/Games/GOG/TEST_DIRECTORY
/mnt/merge/backup/raid-backup/data1/Games/GOG/TEST_DIRECTORY
```
It has been created successfully. Now to check on which branch it's been placed:
```
$ ls -d /mnt/bank/backup1/raid-backup/data1/Games/GOG/TEST_DIRECTORY
ls: cannot access '/mnt/bank/backup1/raid-backup/data1/Games/GOG/TEST_DIRECTORY': No such file or directory

$ ls -d /mnt/bank/backup2/raid-backup/data1/Games/GOG/TEST_DIRECTORY
ls: cannot access '/mnt/bank/backup2/raid-backup/data1/Games/GOG/TEST_DIRECTORY': No such file or directory

$ ls -d /mnt/bank/backup3/raid-backup/data1/Games/GOG/TEST_DIRECTORY
ls: cannot access '/mnt/bank/backup3/raid-backup/data1/Games/GOG/TEST_DIRECTORY': No such file or directory

$ ls -d /mnt/bank/backup4/raid-backup/data1/Games/GOG/TEST_DIRECTORY
/mnt/bank/backup4/raid-backup/data1/Games/GOG/TEST_DIRECTORY
```
So as you can see, `TEST_DIRECTORY` has been placed on the `/mnt/bank/backup4` branch, where its parent `GOG` directory did not exist previously. For good measure, here you can see `/mnt/bank/backup1` has 2.2 TB of free space left, so it's not full at all:
```
$ df -h /mnt/bank/backup1
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda1       7.3T  5.1T  2.2T  70% /mnt/bank/backup1
```


 I would have expected `TEST_DIRECTORY` to be placed under the `/mnt/bank/backup1` branch because its parent `/mnt/bank/backup1/raid-backup/data1/Games/GOG` was already present, while it was not there on any other branch, as you can see from the listings above.

I don't know if I'm misunderstanding mspmfs and this is intended behavior or not. My understanding is that mspmfs would first check for the relative path `raid-backup/data1/Games/GOG` on all branches, find it only on branch 1 and as such `TEST_DIRECTORY` should have been created on branch 1. Clearly however it has been created on branch 4 instead, where `raid-backup/data1/Games/GOG` did not exist. 

If this is correct behavior and I am just misunderstanding mspmfs, what would be the proper settings to get the behavior I expected?

**Reply by trapexit** [reply]
Run the following and give me the output

```
getfattr -n user.mergerfs.func.mkdir /mnt/merge/backup/.mergerfs
getfattr -n user.mergerfs.func.create /mnt/merge/backup/.mergerfs
```

**Reply by red217** [reply]
Here they are:
```
$ getfattr -n user.mergerfs.func.mkdir /mnt/merge/backup/.mergerfs
getfattr: Removing leading '/' from absolute path names
# file: mnt/merge/backup/.mergerfs
user.mergerfs.func.mkdir="mspmfs"

$ getfattr -n user.mergerfs.func.create /mnt/merge/backup/.mergerfs
getfattr: Removing leading '/' from absolute path names
# file: mnt/merge/backup/.mergerfs
user.mergerfs.func.create="mspmfs"
```

**Reply by trapexit** [reply]
Can you please update to a current release? That version is from 2020 and a lot has changed. I don't know what's going on but I don't want to keep debugging such an old release.

**Reply by red217** [reply]
I understand that.
Normally I prefer to to use critical software like things dealing with file storage from the default distro repos to ensure I don't get strange misbehavior with OS updates, but I was already considering trying the current version. I didn't do it before posting here because I wasn't sure whether I simply had it misconfigured or not, but now I take it that the configuration isn't wrong and the behavior is indeed unexpected.

I'll update when I've got a bit of time to mess with it some more and hopefully things will change for the better. Thanks for your time!

**Reply by trapexit** [reply]
You can't trust versions in the distro. They never update them regardless of bugs in the version they originally grab. That's why the first thing I ask in the support section is to check problems against the latest release. I have no control over what is packaged. Only my own releases.

https://github.com/trapexit/mergerfs#support

**Reply by red217** [reply]
Yeah, I see. I'll give the latest release for Debian bullseye a shot.

**Reply by red217** [reply]
I installed mergerfs 2.35.1 and it's behaving as I expected now. Following the same steps described above, the test directory is created on branch 1, so I suppose this was an issue with 2.31.0 and not something configuration related. Everything is working as I want it to now, so thanks for the help!
